File: src/sdk/contract.ts
export const SDK_API_VERSION = 2;

export interface SdkCapabilities {
  structuredEngine: boolean;
  typedFramePlanning: boolean;
  resourceGenerations: boolean;
  graphPlanning: boolean;
  wgslParsing: boolean;
  wgslCompilation: boolean;
  gpuResourcePrimitives: boolean;
  gpuExecution: boolean;
  onnxPrePostprocessing: boolean;
  nativeOnnxSession: boolean;
  browserOnnxSession: boolean;
}

export interface RuntimePublicSurface {
  apiVersion: number;
  methods: string[];
}

export function decodeRuntimePublicSurface(json: string): RuntimePublicSurface {
  const value = JSON.parse(json) as Partial<RuntimePublicSurface>;
  if (
    value.apiVersion !== SDK_API_VERSION
    || !Array.isArray(value.methods)
    || value.methods.some((method) => typeof method !== 'string')
  ) {
    throw new SdkContractError({
      code: 'invalid-response',
      message: 'Rust SDK returned an invalid runtime public surface',
    });
  }
  return value as RuntimePublicSurface;
}

export interface FrameStamp {
  epoch: number;
  frame: number;
  timelineNs: number;
  deadlineNs: number;
}

export interface ContentStamp {
  epoch: number;
  timelineNs: number;
  mediaPtsNs?: number;
}

export interface OutputKey {
  nodeId: string;
  portId: string;
}

export type DeliveryPolicy = 'on-change' | 'latest' | 'every';
export type OutputTransport = 'value' | 'preview' | 'capture' | 'native-present';

export interface OutputSubscription {
  subscriptionId: string;
  output: OutputKey;
  delivery: DeliveryPolicy;
  transport: OutputTransport;
  maxWidth?: number;
  maxHeight?: number;
}

export type OutputPayload =
  | { kind: 'bool'; value: boolean }
  | { kind: 'int' | 'uint' | 'float'; value: number }
  | { kind: 'float-array'; value: number[] }
  | { kind: 'json'; value: unknown }
  | { kind: 'resource'; value: { handle: number } }
  | { kind: 'tensor'; value: { handle: number; dtype: string; shape: number[] } }
  | { kind: 'bytes'; value: number[] };

export interface OutputState {
  output: OutputKey;
  graphRevision: number;
  outputGeneration: number;
  evaluationStamp: FrameStamp;
  contentStamp: ContentStamp;
  payload: OutputPayload;
}

export interface OutputDelivery {
  subscriptionId: string;
  state: OutputState;
}

export interface SubscriptionInvalidation {
  subscriptionId: string;
  output: OutputKey;
  reason: string;
}

export interface OutputDeliveryBatch {
  frameStamp?: FrameStamp;
  deliveries: OutputDelivery[];
  invalidations: SubscriptionInvalidation[];
}

export type SdkErrorCode =
  | 'disposed'
  | 'invalid-frame'
  | 'invalid-graph'
  | 'invalid-resource'
  | 'invalid-state'
  | 'not-prepared'
  | 'stale-revision'
  | 'unknown-node'
  | 'protocol-mismatch'
  | 'invalid-response';

export interface SdkErrorPayload {
  code: SdkErrorCode;
  message: string;
  nodeId?: string;
  details?: string;
}

export type EngineState = 'empty' | 'ready' | 'running' | 'paused' | 'stopped' | 'disposed';

export type EngineEvent =
  | { type: 'state'; state: EngineState }
  | { type: 'graph-ready'; revision: number }
  | { type: 'resource-invalidated'; nodeId: string; generation: number }
  | { type: 'resource-released'; nodeId: string; generation: number }
  | {
      type: 'frame-planned';
      frame: number;
      revision: number;
      commandCount: number;
      dirtyNodeCount: number;
    };

export class SdkContractError extends Error {
  readonly code: SdkErrorCode;
  readonly nodeId?: string;
  readonly details?: string;

  constructor(payload: SdkErrorPayload) {
    super(payload.message);
    this.name = 'SdkContractError';
    this.code = payload.code;
    this.nodeId = payload.nodeId;
    this.details = payload.details;
  }
}

export function decodeSdkError(error: unknown): SdkContractError {
  if (error instanceof SdkContractError) return error;
  const source = error instanceof Error ? error.message : error;
  if (typeof source === 'string') {
    try {
      const parsed = JSON.parse(source) as Partial<SdkErrorPayload>;
      if (typeof parsed.code === 'string' && typeof parsed.message === 'string') {
        return new SdkContractError(parsed as SdkErrorPayload);
      }
    } catch {
      return new SdkContractError({ code: 'invalid-response', message: source });
    }
    return new SdkContractError({ code: 'invalid-response', message: source });
  }
  return new SdkContractError({
    code: 'invalid-response',
    message: 'Rust SDK returned a non-structured error',
    details: String(source),
  });
}

export function decodeCapabilities(json: string): SdkCapabilities {
  const value = JSON.parse(json) as Partial<SdkCapabilities>;
  const keys: ReadonlyArray<keyof SdkCapabilities> = [
    'typedFramePlanning',
    'resourceGenerations',
    'structuredEngine',
    'graphPlanning',
    'wgslParsing',
    'wgslCompilation',
    'gpuResourcePrimitives',
    'gpuExecution',
    'onnxPrePostprocessing',
    'nativeOnnxSession',
    'browserOnnxSession',
  ];
  if (!keys.every((key) => typeof value[key] === 'boolean')) {
    throw new SdkContractError({
      code: 'invalid-response',
      message: 'Rust SDK returned invalid capabilities',
    });
  }
  return value as SdkCapabilities;
}

export function decodeEngineEvents(json: string): EngineEvent[] {
  const value: unknown = JSON.parse(json);
  if (!Array.isArray(value)) {
    throw new SdkContractError({
      code: 'invalid-response',
      message: 'Rust SDK returned invalid engine events',
    });
  }
  return value as EngineEvent[];
}

File: crates/open_quartz/src/runtime/contract.rs
use serde::{Deserialize, Serialize};

use crate::ffi::SDK_API_VERSION;

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FrameStamp {
    pub epoch: u64,
    pub frame: u64,
    pub timeline_ns: u64,
    pub deadline_ns: u64,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ContentStamp {
    pub epoch: u64,
    pub timeline_ns: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub media_pts_ns: Option<i64>,
}

#[derive(Clone, Debug, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputKey {
    pub node_id: String,
    pub port_id: String,
}

impl OutputKey {
    pub fn new(node_id: impl Into<String>, port_id: impl Into<String>) -> Self {
        Self {
            node_id: node_id.into(),
            port_id: port_id.into(),
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum DeliveryPolicy {
    OnChange,
    Latest,
    Every,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum OutputTransport {
    Value,
    Preview,
    Capture,
    NativePresent,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", content = "value", rename_all = "kebab-case")]
pub enum OutputPayload {
    Bool(bool),
    Int(i64),
    Uint(u64),
    Float(f64),
    FloatArray(Vec<f64>),
    Json(serde_json::Value),
    Resource {
        handle: u64,
    },
    Tensor {
        handle: u64,
        dtype: String,
        shape: Vec<u64>,
    },
    Bytes(Vec<u8>),
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputState {
    pub output: OutputKey,
    pub graph_revision: u32,
    pub output_generation: u64,
    pub evaluation_stamp: FrameStamp,
    pub content_stamp: ContentStamp,
    pub payload: OutputPayload,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputSubscription {
    pub subscription_id: String,
    pub output: OutputKey,
    pub delivery: DeliveryPolicy,
    pub transport: OutputTransport,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_width: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub max_height: Option<u32>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputDelivery {
    pub subscription_id: String,
    pub state: OutputState,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubscriptionInvalidation {
    pub subscription_id: String,
    pub output: OutputKey,
    pub reason: String,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OutputDeliveryBatch {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frame_stamp: Option<FrameStamp>,
    pub deliveries: Vec<OutputDelivery>,
    pub invalidations: Vec<SubscriptionInvalidation>,
}

#[derive(Clone, Copy, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Viewport {
    pub x: f32,
    pub y: f32,
    pub width: f32,
    pub height: f32,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum PresentationFit {
    Contain,
    Cover,
    Stretch,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PresentationItem {
    pub output: OutputKey,
    pub resource_handle: u64,
    pub viewport: Viewport,
    pub fit: PresentationFit,
    pub z_index: i32,
    pub evaluation_stamp: FrameStamp,
    pub content_stamp: ContentStamp,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PresentationSet {
    pub group_id: String,
    pub frame_stamp: FrameStamp,
    pub items: Vec<PresentationItem>,
}

#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AsyncCompletionEnvelope {
    pub node_id: String,
    pub graph_revision: u32,
    pub node_generation: u32,
    pub input_stamp: FrameStamp,
    pub content_stamp: ContentStamp,
    pub outputs: Vec<(OutputKey, OutputPayload)>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub enum DataPathMode {
    #[serde(rename = "cpu-copy")]
    CpuCopy,
    #[serde(rename = "external-frame/no-cpu-readback")]
    ExternalFrameNoCpuReadback,
    #[serde(rename = "shared-gpu")]
    SharedGpu,
    #[serde(rename = "native-present")]
    NativePresent,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct RuntimeCapabilities {
    pub data_paths: Vec<DataPathMode>,
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PublicSurfaceManifest {
    pub api_version: u32,
    pub methods: &'static [&'static str],
}

pub fn public_surface_manifest() -> PublicSurfaceManifest {
    PublicSurfaceManifest {
        api_version: SDK_API_VERSION,
        methods: &[
            "set_graph",
            "register_resource",
            "remove_resource",
            "play",
            "advance",
            "subscribe_output",
            "update_output_subscription",
            "unsubscribe_output",
            "subscribe_presentation",
            "update_presentation",
            "unsubscribe_presentation",
            "submit_completion",
            "drain_work",
            "drain_deliveries",
            "drain_events",
            "capabilities",
            "pause",
            "resume",
            "stop",
            "dispose",
        ],
    }
}
